12.0 automatic credit validation

stephansainleger opened the pull request.

Commit:
[CLN] lcc_comchain_base: black formating

Commit:
[IMP] lcc_comchain_base: add automatic top up process

Commit:
[IMP] lcc_lokavaluto_app_connection: add wallet search quick filters

Commit:
[CLN] lcc_lokavaluto_app_connection: remove useless file

Commit:
[CLN] lcc_lokavaluto_app_connection: rename res partner view file

Commit:
[IMP] lcc_lokavaluto_app_connection: new access rights on wallets

Commit:
[IMP] lcc_comchain_base: new access rights on wallets

Commit:
[FIX] lcc_comchain_base: allow _compute_status() on multi records

Commit:
[CLN] lcc_lokavaluto_app_connection: remove useless dependency

on lcc_members_qr add-on

Commit:
[CLN] lcc_lokavaluto_app_connection: remove useless field and code

field in_mobile_app is useless today

Commit:
[CLN] lcc_lokavaluto_app_connection: remove useless services

njeudy reviewed (dismissed):
- Code review et check with stephan - no functionnal test

Commit:
[IMP] lcc_lokavaluto_app_connection: add partner in backend form view

vaab commented on lcc_comchain_base/models/account_invoice.py:
```diff
@@ -1,19 +1,21 @@
 
     comchain_amount_credited = fields.Float(
-        string="Comchain amount credited", track_visibility="always"
+        string="Comchain amount credited", track_visibility="always", store=True
```
useless

vaab commented on lcc_comchain_base/models/account_invoice.py:
```diff
@@ -1,19 +1,21 @@
     )
     comchain_amount_to_credit = fields.Float(
-        string="Comchain amount to credit", track_visibility="always"
+        string="Comchain amount to credit", track_visibility="always", store=True
```
useless

vaab commented on lcc_comchain_base/models/account_invoice.py:
```diff
@@ -1,19 +1,21 @@
-        string="Comchain amount to credit", track_visibility="always"
+        string="Comchain amount to credit", track_visibility="always", store=True
     )
+    comchain_transaction_number = fields.Char('Comchain Transaction Number', track_visibility="always", store=True)
```
remove `store=True` (only useful for computed values)

vaab replied:
I would suggest for consistency that you either put 'string=' either remove it on the 2 previous line. (did you use `black` by the way on this commit, the formatting is quite different between the 2 previous line and this one ?)

vaab commented on lcc_comchain_base/models/account_invoice.py:
```diff
@@ -35,6 +37,33 @@ def action_invoice_paid(self):
                     }
                 )
+                if self.company_id.activate_automatic_topup:
+                    odoo_wallet = Pyc3l().Wallet.from_json(self.company_id.odoo_wallet_partner_id.lcc_backend_ids[0].comchain_wallet)
+                    partner_wallet = invoice.partner_id._comchain_backend()
```
This line is quite costly, `Pyc3l()` will elect an endpoint, and this should probably not be done for each invoice at least. 
Also, you probably want to not run all this if total amount is 0... which I guess probably couldn't happen for not so obvious reason, so either a comment explaining why it is sure that amount has to be > 0, either add a defensive instruction here. 
The tests that follow (and might exit with exceptions) should probably be before the loading of the Wallet.

vaab replied:
Another concern for me is the usage of this very long resolution line:

self.company_id.odoo_wallet_partner_id... 

A lot of part of these are used again in the following code, why not put a variable out. 
And second, you are making a lot of assumption that all elements exists on the way... which might be okay, but this is not clear to me for the moment.

vaab commented on lcc_comchain_base/models/account_invoice.py:
```diff
@@ -35,6 +37,33 @@ def action_invoice_paid(self):
+                        message_from=self.company_id.message_from,
+                        message_to=self.company_id.message_to,
+                    )
+                    if res != "":
```
I'm surprised of this test, can `res` be something else than the transaction hash ? What makes you think that ?
And if it did... that's an indication of an important error. Should'nt we fail (raise an exception) instead of continuing silently ?
I was wondering also if we should double check that everything is alright by asking comchain if he knows about the given transaction... and we could even check the amount and the recipient.

vaab commented on lcc_comchain_base/models/company.py:
```diff
@@ -1,7 +1,18 @@
 
+    @api.model
+    def _default_messages(self):
+        message = "Top-up from " + str(self.name)
```
the combination of `@api.model` and `self.name` seems wrong to me here (you probably can safely remove `@api.model`). Not sure also why you need this variable `message`, and `str(..)` of a string is not necessary.

vaab commented on lcc_comchain_base/models/company.py:
```diff
@@ -1,7 +1,18 @@
+        return message
+
     comchain_currency_name = fields.Char(string="Currency name")
+    activate_automatic_topup = fields.Boolean('Activate Automatic Topup')
+    safe_wallet_partner_id = fields.Many2one('res.partner', string="Safe Wallet Partner")
+    odoo_wallet_partner_id = fields.Many2one('res.partner', string="Odoo Wallet Partner")
+    comchain_odoo_wallet_password = fields.Char(string="Odoo wallet password")
```
why do you use `string=` sometimes and other times you don't. As the first keyword, I think you can remove it everywhere. (except Many2one, Selection...)... or you could write it everywhere :-)

vaab replied:
for me, a lot of these should be on `lcc_currency` or something like that. It only concern one currency and actually, probably only comchain currency. But I understand that you put it here temporarily.

vaab commented on lcc_comchain_base/views/account_invoice_view.xml:
```diff
@@ -14,8 +14,14 @@
                     credit request. </div>
             </xpath>
+            <xpath expr="//page[@name='other_info']/group" position="inside">
+                <group name="credit_data" string="Monnaie numérique">
+                    <field name="comchain_amount_to_credit" string="A créditer" readonly="True" />
+                    <field name="comchain_amount_credited" string="Crédité" readonly="True" />
```
is that okay to use french here ? And BTW, thank you so much for these debugging info !.

vaab commented on lcc_comchain_base/models/account_invoice.py:
```diff
@@ -35,6 +37,33 @@ def action_invoice_paid(self):
+                            {
+                                "comchain_amount_to_credit": 0,
+                                "comchain_amount_credited": amount,
+                                "comchain_transaction_number": res
```
That's excellent ! we need to link the transaction number from the financial backend to the invoice ! Thanks !

vaab commented on lcc_comchain_base/models/account_invoice.py:
```diff
@@ -35,6 +37,33 @@ def action_invoice_paid(self):
+                    if res != "":
+                        invoice.write(
+                            {
+                                "comchain_amount_to_credit": 0,
```
it is probably not the best of idea to re-write this value (a second time).

vaab commented on lcc_lokavaluto_app_connection/views/res_partner.xml:
```diff
@@ -7,8 +7,10 @@
+                <button name="in_mobile_app_button" type="object" class="oe_stat_button"
+                    icon="fa-map-pin" attrs="{'invisible': [('is_company','=',False)]}">
+                    <field name="in_mobile_app" widget="boolean_button"
+                        options='{
```
only space changes ? if yes it would have been nice in another commit.

vaab commented on lcc_lokavaluto_app_connection/security/backends_security.xml:
```diff
@@ -0,0 +1,13 @@
+        </record>
+
+        <record id="group_wallet_full_manager" model="res.groups">
+            <field name="name">Wallet_full_manager</field>
```
are the underscore normal here ?

vaab commented on lcc_lokavaluto_app_connection/security/backends_security.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data noupdate="0">
```
isn't that the default ?

vaab commented on lcc_lokavaluto_app_connection/models/res_partner.py:
```diff
@@ -14,8 +14,20 @@ class ResPartner(models.Model):
+    @api.depends("lcc_backend_ids")
+    def _compute_nb_wallets(self):
+        self.nb_wallets = len(self.lcc_backend_ids)
+        self.nb_wallets_to_confirm = len(self.lcc_backend_ids.filtered(lambda x: x.status == "to_confirm"))
+        self.nb_wallets_inactive = len(self.lcc_backend_ids.filtered(lambda x: x.status == "inactive"))
+        self.nb_wallets_blocked = len(self.lcc_backend_ids.filtered(lambda x: x.status == "blocked"))
```
it's a little bit sad to parse `lcc_backend_ids` 3 times to separate them in three parts.

Commit:
[CLN] Add .gitignore for all repo add-ons

Commit:
[IMP] lcc_lokavaluto_app_connection: add active field in wallet

Commit:
[CLN] lcc_comchain_base: move res.partner.backend in dedicated file

Commit:
[IMP] lcc_comchain_base: standardize top up process

makes comchain top up process/code organization similar Cyclos one.

Commit:
[IMP] Refactoring: create get_wallet() to return partner wallet

add-ons concerned: lcc_lokavaluto_app_connection, lcc_cyclos_base,
lcc_comchain_base
Removes function _comchain_backend() and _cyclos_backend() to replace
them by get_wallet() in lcc_lokavaluto_app_connection

Commit:
[IMP] global: add credit.request for numeric lcc, global refactoring

Commit:
[IMP] lcc_lokavaluto_app_connection: add new credit request wizard

available from res.partner.backend form view

Commit:
[IMP] lcc_lokavaluto_app_connection: prevent wallet delete

Commit:
[IMP] lcc_lokavaluto_app_connection: add "Try again" button

available on credit request with state in error.

Commit:
[FIX] global: transform get_wallet() in get_wallets()

and refactor concerned function to support wallet list

Commit:
[IMP] lcc_comchain_base: refactor comchain wallet validation

Commit:
[IMP] global: adapt pending-topup service to recent refactoring

makes pending-topup service work with any financial backend

Commit:
[IMP] lcc_cyclos_base: add get_cyclos_server_domain function

Commit:
[FIX] global: adapt backend_key name translation to global wallet list

Commit:
[IMP] lcc_lokavaluto_app_connection: update translations

Commit:
[IMP] lcc_cyclos_base: update translations

Commit:
[IMP] lcc_comchain_base: update translations

Commit:
[NEW] lcc_credit_requests_from_contracts: create add-on

Commit:
[CLN] lcc_cyclos_base: fix typo

Commit:
[CLN] lcc_cyclos_base: fix typo in docstring

Commit:
[IMP] lcc_cyclos_base: new access rights on wallets

Commit:
[IMP] lcc_lokavaluto_app_connection: refactor credit request data

Refactors credit request retrieval process in the API. There was no
need of managing the data sent through the API in credit request or
wallet models.

Commit:
[IMP] global: refactor wallet data retrieval processes

Refactors wallet data retrieval processes so that the data
normalization for the API is not managed by wallet and partner models anymore.
